Create geoip plugin (#4688)
miekg/coredns: plugin/geoip/README.md

# geoip

## Name
*geoip* - Lookup maxmind geoip2 databases using the client IP, then add associated geoip data to the context request.

## Description
The *geoip* plugin add geo location data associated with the client IP, it allows you to configure a [geoIP2 maxmind database](https://dev.maxmind.com/geoip/docs/databases) to add the geo location data associated with the IP address.

> The data is added leveraging the [metadata](../metadata/README.md) plugin, values can then be retrieved using it as well, for example:

Reviewer: ```suggestion
The data is added leveraging the *metadata* plugin, values can then be retrieved using it as well, for example:
```

```go
import (
    "strconv"
    "github.com/coredns/coredns/plugin/metadata"
)
// ...
if getLongitude := metadata.ValueFunc(ctx, "geoip/longitude"); getLongitude != nil {
    if longitude, err := strconv.ParseFloat(getLongitude(), 64); err == nil {
		// Do something useful with longitude.
	}
} else {
    // The metadata label geoip/longitude for some reason, was not set.
}
// ...
```

## Databases

> Right now only databases providing the city schema such as `City` and `Enterprise` are supported, other databases types with different schemas are not supported yet. 

Reviewer: ```suggestion
The supported databases use city schema such as `City` and `Enterprise` . Other databases types with different schemas are not supported yet. 
```

> You can download a free and public City database [here](https://www.maxmind.com/en/geoip2-city).

Reviewer: ```suggestion
You can download a [free and public City database](https://www.maxmind.com/en/geoip2-city).
```

## Syntax

> * **DBFILE** the mmdb database file path.

Reviewer: move the DBFILE bullet below the syntax example

Author: Done

```txt
geoip [DBFILE]
```

> You can specifiy a list of languages for `names` fields.

Reviewer: the names field is a thing in the geo IP db? Prolly worth explaining a tiny bit more, or remove this entirely and explain it below in the bullet item

Author: Deferring action until discussion around languages is clear, because we may end up removing languages option which would get this text removed entirely.

Author: Removed.

```txt
geoip [DBFILE] {
    languages [CODE]...
}
```

* `languages` configures a list of languages for which you want to create a label, by default it's set to English: `en`, note that except for English many languages may not be available in the database, language must be available in your geoip2 database, a list of the codes supported can be found in [Locale Codes](#LocaleCodes).

## Examples
The following configuration configures the `City` database adding names in English, French, Spanish and Simplified Chinese.
```txt
. {
    geoip /opt/geoip2/db/GeoLite2-City.mmdb {

>         languages en fr es zh-CN

Reviewer: looking at the use of LANG in the labels and I don't feel it's good enough. You're basically forcing the user of this data to looks for all possible LANGs and there is a coupling the corefile and the code.
Why is a language even needed? I think we can default to english and drop the whole LANG option

Author: My initial thought was that some users may have a use case for getting "names", such as city names, country names, etc. in a different language than English, and giving them a a good default (English) while still letting them change that choice could be useful.

I'm happy to remove the option and provide English only, and do something else in the future when a clear use case appears.

Author: Removed languages option, and default to English only.

    }
    metadata # Note that metadata plugin must be enabled as well.
}
```

## Metadatada Labels
A limited set of fields will be exported as labels, all values are stored using strings **regardless of their underlying value type**, and therefore you may have to convert it back to its original type, note that numeric values are always represented in base 10.

> **LANG** Language location ISO??? code, if database does not have a name for the selected language the label

Reviewer: ```suggestion
**LANG** is the language  ISO code, if database does not have a name for the selected language the label
```
